File: BMI_Calculator.c
#include <math.h>
#include <stdbool.h>
#include <stdio.h>

bool calculate_bmi(double mass_kg, double height_m, double *result)
{
    if (result == NULL || !isfinite(mass_kg) || !isfinite(height_m) ||
        mass_kg <= 0.0 || height_m <= 0.0) {
        return false;
    }
    *result = mass_kg / (height_m * height_m);
    return true;
}

const char *bmi_category(double bmi)
{
    if (!isfinite(bmi) || bmi < 0.0) {
        return "Invalid";
    }
    if (bmi < 18.5) {
        return "Underweight";
    }
    if (bmi < 25.0) {
        return "Normal weight";
    }
    if (bmi < 30.0) {
        return "Overweight";
    }
    return "Obese";
}

#ifndef BMI_CALCULATOR_NO_MAIN
int main(void)
{
    double mass;
    double height;
    double bmi;

    printf("Mass (kg): ");
    if (scanf("%lf", &mass) != 1) {
        fprintf(stderr, "Error: mass must be a number.\n");
        return 1;
    }

    printf("Height (m): ");
    if (scanf("%lf", &height) != 1) {
        fprintf(stderr, "Error: height must be a number.\n");
        return 1;
    }

    if (!calculate_bmi(mass, height, &bmi)) {
        fprintf(stderr, "Error: mass and height must be finite and greater than zero.\n");
        return 1;
    }

    printf("BMI: %.2f\n", bmi);
    printf("Category: %s\n", bmi_category(bmi));
    return 0;
}
#endif

File: BMI_Calculator.py
"""Body mass index calculator with input validation."""

from __future__ import annotations

import math


def calculate_bmi(mass_kg: float, height_m: float) -> float:
    """Calculate BMI from mass in kilograms and height in metres."""

    mass_kg = float(mass_kg)
    height_m = float(height_m)
    if not math.isfinite(mass_kg) or not math.isfinite(height_m):
        raise ValueError("Mass and height must be finite numbers.")
    if mass_kg <= 0 or height_m <= 0:
        raise ValueError("Mass and height must be greater than zero.")
    return mass_kg / height_m**2


def bmi_category(bmi: float) -> str:
    """Return the standard adult BMI category for a BMI value."""

    if not math.isfinite(bmi) or bmi < 0:
        raise ValueError("BMI must be a finite, non-negative number.")
    if bmi < 18.5:
        return "Underweight"
    if bmi < 25:
        return "Normal weight"
    if bmi < 30:
        return "Overweight"
    return "Obese"


def main() -> int:
    try:
        mass = float(input("Mass (kg): ").strip())
        height = float(input("Height (m): ").strip())
        bmi = calculate_bmi(mass, height)
    except ValueError as error:
        print(f"Error: {error}")
        return 1

    print(f"BMI: {bmi:.2f}")
    print(f"Category: {bmi_category(bmi)}")
    return 0


if __name__ == "__main__":
    raise SystemExit(main())
